internal/cmd/controller/clustermonitor/monitor.go

```diff
@@ -0,0 +1,132 @@
+		threshold := time.Duration(cfg.AgentCheckinInterval.Seconds() * 1.1)
+
+		UpdateOfflineBundleDeployments(ctx, c, threshold)
+	}
+}
+
+func UpdateOfflineBundleDeployments(ctx context.Context, c client.Client, threshold time.Duration) {
+	logger := ctrl.Log.WithName("cluster status monitor")
+
+	clusters := &v1alpha1.ClusterList{}
+	if err := c.List(ctx, clusters); err != nil {
+		logger.Error(err, "Failed to get list of clusters")
+		return
+	}
+
+	for _, cluster := range clusters.Items {
+		lastSeen := cluster.Status.Agent.LastSeen
+
+		logger.Info("Checking cluster status", "cluster", cluster.Name, "last seen", lastSeen.UTC().String())
+
+		// XXX: do we want to run this more than once per cluster, updating the timestamp each time?
+		// Or would it make sense to keep the oldest possible timestamp in place, for users to know since when the
+		// cluster is offline?
+
+		// lastSeen being 0 would typically mean that the cluster is not registered yet, in which case bundle
+		// deployments should not be deployed there.
+		if lastSeen.IsZero() || time.Now().UTC().Sub(lastSeen.UTC()) < threshold {
+			continue
+		}
+
+		logger.Info("Detected offline cluster", "cluster", cluster.Name)
+
+		// Cluster is offline
+		bundleDeployments := &v1alpha1.BundleDeploymentList{}
+		if err := c.List(ctx, bundleDeployments, client.InNamespace(cluster.Status.Namespace)); err != nil {
+			logger.Error(
+				err,
+				"Failed to get list of bundle deployments for offline cluster",
+				"cluster",
+				cluster.Name,
+				"namespace",
+				cluster.Status.Namespace,
+			)
+			continue
+		}
+
+		for _, bd := range bundleDeployments.Items {
+			logger.Info("Updating bundle deployment in offline cluster", "cluster", cluster.Name, "bundledeployment", bd.Name)
+			err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				t := &v1alpha1.BundleDeployment{}
+				nsn := types.NamespacedName{Name: bd.Name, Namespace: bd.Namespace}
+				if err := c.Get(ctx, nsn, t); err != nil {
+					return err
+				}
+				t.Status = bd.Status
+				// Any information about resources living in an offline cluster is likely to be
+				// outdated.
+				t.Status.ModifiedStatus = nil
+				t.Status.NonReadyStatus = nil
+
+				for _, cond := range bd.Status.Conditions {
+					switch cond.Type {
+					// XXX: which messages do we want to set and where?
+					case "Ready":
+						// FIXME: avoid relying on agent pkg for this?
+						mc := monitor.Cond(v1alpha1.BundleDeploymentConditionReady)
+						mc.SetError(&bd.Status, "Cluster offline", fmt.Errorf("cluster is offline"))
+						// XXX: do we want to set Deployed and Installed conditions as well?
+						// XXX: should we set conditions to `Unknown`?
+					case "Monitored":
+						mc := monitor.Cond(v1alpha1.BundleDeploymentConditionMonitored)
+						mc.SetError(&bd.Status, "Cluster offline", fmt.Errorf("cluster is offline"))
+
+					}
+				}
+
+				return c.Status().Update(ctx, t)
```

Comment:
Shouldn't it be setting the Error on `t` instead of `bd`?
The update call is done with `t`.

Reply:
Nice catch, thanks! This still behaved as expected because of references, but was clearly not ideal. Fixed.